Shader compilation error occurs when writing to `FOG` in a custom shader that uses `render_mode fog_disabled`

### Tested versions

- Reproducible in: 4.2.stable, 4.2.2.stable, 4.3.beta da4f6e439

### System information

Godot v4.3.beta (da4f6e439) - Fedora Linux 40 (KDE Plasma) - X11 - GLES3 (Compatibility) - NVIDIA GeForce RTX 4090 (nvidia; 555.58.02) - 13th Gen Intel(R) Core(TM) i9-13900K (32 Threads)

### Issue description

A shader compilation error occurs when writing to `FOG` in a custom shader that uses `render_mode fog_disabled`. This can be reproduced with all rendering methods.

I suggest ignoring the writes to `FOG` in this case (likely better for shaders that can have their render mode adjusting for a script). We could also forbid writing to `FOG` with an error message with this case, but it may be more intrusive.

### Forward+

```
   at: _compile_variant (servers/rendering/renderer_rd/shader_rd.cpp:297)
ERROR: Error compiling Fragment shader, variant #9 ().
   at: _compile_variant (servers/rendering/renderer_rd/shader_rd.cpp:293)
ERROR: Failed parse:
WARNING: 0:74: '' : all default precisions are highp; use precision statements to quiet warning, e.g.:
         "precision mediump int; precision highp float;" 
ERROR: 0:2242: 'fog' : undeclared identifier 
ERROR: 0:2242: 'a' : vector swizzle selection out of range 
ERROR: 0:2242: '' : compilation terminated 
ERROR: 3 compilation errors.  No code generated.

   at: _compile_variant (servers/rendering/renderer_rd/shader_rd.cpp:294)
ERROR: code:

[... Shader code printed here ...]

   at: _compile_variant (servers/rendering/renderer_rd/shader_rd.cpp:297)
ERROR: Parameter "p_version->variants" is null.
   at: _allocate_placeholders (servers/rendering/renderer_rd/shader_rd.cpp:487)
ERROR: Parameter "p_version->variants" is null.
   at: _allocate_placeholders (servers/rendering/renderer_rd/shader_rd.cpp:487)
ERROR: Parameter "p_version->variants" is null.
   at: _allocate_placeholders (servers/rendering/renderer_rd/shader_rd.cpp:487)
ERROR: Condition "!shader_singleton->shader.version_is_valid(version)" is true.
   at: set_code (servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp:182)
```

### Mobile

```
ERROR: Error compiling Fragment shader, variant #0 ().
   at: _compile_variant (servers/rendering/renderer_rd/shader_rd.cpp:293)
ERROR: Failed parse:
WARNING: 0:69: '' : all default precisions are highp; use precision statements to quiet warning, e.g.:
         "precision mediump int; precision highp float;" 
ERROR: 0:1788: 'fog' : undeclared identifier 
ERROR: 0:1788: 'a' : vector swizzle selection out of range 
ERROR: 0:1788: '' : compilation terminated 
ERROR: 3 compilation errors.  No code generated.

   at: _compile_variant (servers/rendering/renderer_rd/shader_rd.cpp:294)

[... Shader code printed here ...]

   at: _compile_variant (servers/rendering/renderer_rd/shader_rd.cpp:297)
ERROR: Condition "!shader_singleton->shader.version_is_valid(version)" is true.
   at: set_code (servers/rendering/renderer_rd/forward_mobile/scene_shader_forward_mobile.cpp:193)
```

### Compatibility

```
WARNING: shader failed to compile, unable to bind shader.
     at: _version_bind_shader (./drivers/gles3/shader_gles3.h:222)

[... Shader code printed here ...]

ERROR: SceneShaderGLES3: Fragment shader compilation failed:
0(1198) : error C1503: undefined variable "fog"
0(1199) : error C1503: undefined variable "fog"

   at: _display_error_with_code (drivers/gles3/shader_gles3.cpp:254)
ERROR: Method/function failed.
   at: _compile_specialization (drivers/gles3/shader_gles3.cpp:396)
```

### Steps to reproduce

- Write to `FOG` in a custom shader that has `render_mode fog_disabled;` at its top.

### Minimal reproduction project (MRP)

[Fog.Shader.Compatibility.Issue.MRP2.zip](https://github.com/user-attachments/files/16322950/Fog.Shader.Compatibility.Issue.MRP2.zip)

--- a/drivers/gles3/shaders/scene.glsl
+++ b/drivers/gles3/shaders/scene.glsl
@@ -2002,9 +2002,7 @@ void main() {
 #ifdef PREMUL_ALPHA_USED
 	float premul_alpha = 1.0;
 #endif // PREMUL_ALPHA_USED
-#ifndef FOG_DISABLED
 	vec4 fog = vec4(0.0);
-#endif // !FOG_DISABLED
 #if defined(CUSTOM_RADIANCE_USED)
 	vec4 custom_radiance = vec4(0.0);
 #endif

--- a/servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl
+++ b/servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl
@@ -1225,9 +1225,7 @@ void fragment_shader(in SceneData scene_data) {
 	float anisotropy = 0.0;
 	vec2 anisotropy_flow = vec2(1.0, 0.0);
 	vec3 energy_compensation = vec3(1.0);
-#ifndef FOG_DISABLED
 	vec4 fog = vec4(0.0, 0.0, 0.0, 1.0);
-#endif // !FOG_DISABLED
 #if defined(CUSTOM_RADIANCE_USED)
 	vec4 custom_radiance = vec4(0.0);
 #endif

--- a/servers/rendering/renderer_rd/shaders/forward_mobile/scene_forward_mobile.glsl
+++ b/servers/rendering/renderer_rd/shaders/forward_mobile/scene_forward_mobile.glsl
@@ -1143,9 +1143,7 @@ void main() {
 #ifdef PREMUL_ALPHA_USED
 	float premul_alpha_highp = 1.0;
 #endif
-#ifndef FOG_DISABLED
 	vec4 fog_highp = vec4(0.0);
-#endif // !FOG_DISABLED
 #if defined(CUSTOM_RADIANCE_USED)
 	vec4 custom_radiance_highp = vec4(0.0);
 #endif
@@ -1284,9 +1282,7 @@ void main() {
 #ifdef PREMUL_ALPHA_USED
 	half premul_alpha = half(premul_alpha_highp);
 #endif
-#ifndef FOG_DISABLED
 	hvec4 fog = hvec4(fog_highp);
-#endif
 #ifdef CUSTOM_RADIANCE_USED
 	hvec4 custom_radiance = hvec4(custom_radiance_highp);
 #endif

